---
title: "Synthase expression time course"
output: html_notebook
---

# Synthase expression time course
I want to make a plot where time in weeks is on the  x axis and pfam00765 presence or absence is on the y axis. I will use the metatranscriptomes from each week aligned to the 31/32 and Acidovorax. 

Go to IMG/JGI, then OMICS > RNAseq, search for "Lake Washington" and select the number link under the "Total RNA seq data sets tab" in the row corresponding to the study: "Freshwater sediment methanotrophic microbial communities from Lake Washington under simulated oxygen tension" (will be 963 or so studies).

I am only interested in the metatranscriptome data sets aligned to either Acidovorax or M. tundripaludum 31/32, so I will sort all the data sets by filtering the column based on one of these reference data sets. Then click "Select All" and go to the "Multiple sample analysis" tab in gray on the same page. Then keep all of the standard parameters and click "Gene expression summary". Then "Select All" rows in the resulting table and Export. Copy and paste the results from the downloaded text file into excel to save as a csv or xlxs file. Open that file.

Search for "homoserine lactone" and in the "Gene product" column. Save the rows that correspond to these genes and save them in a new xlxs and transpose them so study names are the row names and coverage is the value. The experimental data is saved in the long title names for each row and is not easy to sort, so we will make new columns in our table that pulls out the metadata for each sample. 

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
```
```{r}
# for editing theme
my_theme <- theme_bw() +
        theme(text=element_text(size = 12),
        axis.text.x=element_text(size=12, color = "black"), 
        axis.text.y = element_text(size=12, color = "black"),
        title = element_text(size = 12, color = "black"), 
        panel.grid.minor = element_blank(), 
        element_line(linewidth = 1, colour = "black")
                )
```


```{r}
# Load the data
coverage <- read.csv("31-32_MTX_pfam0765_coverage.csv")

# generate nicknames
generateNickname <- function(name) {
  nickname <- NA_character_ # Default to NA character vector
  
  if(str_detect(name, "Sediment Metatranscriptome")) {
    matches <- str_match(name, "Sediment Metatranscriptome (\\d+)_([A-Z0-9]+)")
    nickname <- paste0("MT", matches[,2], "_", matches[,3])
    if(str_detect(name, "Raw Count")) nickname <- paste0(nickname, "_rc")
  } else {
    nickname <- name # Default case
  }
  
  # Replace dashes and spaces with underscores for all nicknames
  nickname <- str_replace_all(nickname, "[- ]", "_")
  
  return(nickname)
}

# Then apply this function to your dataframe
coverage$Nickname <- sapply(coverage$sample, generateNickname)

# Add new columns: "sample_ID", "starting_oxygen", "week", "raw_count"
coverage <- coverage %>%
  mutate(
    sample_ID = as.numeric(str_extract(Nickname, "(?<=(MT))\\d+")), 
    starting_oxygen = case_when(
      str_detect(Nickname, "_LOW") ~ "low",
      str_detect(Nickname, "_HOW") ~ "high",
      TRUE ~ NA_character_), 
    week = as.numeric(str_extract(Nickname, "(?<=_(LOW|HOW))\\d+")), 
    raw_count = if_else(str_detect(Nickname, "_rc$"), "raw count", "normalized coverage")
    )

# Save the modified dataframe
write.csv(coverage, "31-32_MTX_pfam0765_coverage_metadata.csv", row.names = FALSE)
```
Okay now my dataframe looks good and I can organize everything by the experimental details. 
Let's make a plot of the normalized coverage of pfam0765 (homoserine lactone synthetase) over time for metatranscriptome datasets aligned to the 31/32 genome. 

```{r}
ggplot(coverage, aes(x=week, y=coverage)) +
  geom_line() + 
  xlab("")

# separate raw couts from normalized coverage

raw <- coverage %>% 
  filter(grepl("raw count", raw_count, fixed = TRUE))

norm <- coverage %>% 
  filter(grepl("normalized coverage", raw_count, fixed = TRUE))

# within the normalized coverage data, separate high and low oxygen conditions
HOW <- norm %>% 
  filter(grepl("high", starting_oxygen, fixed = TRUE))

LOW <- norm %>% 
  filter(grepl("low", starting_oxygen, fixed = TRUE))

# Calculate averages for each week for the HOW data
HOW_avg <- HOW %>%
  group_by(week) %>%
  summarise(average_coverage = mean(coverage, na.rm = TRUE)) # Compute mean, ignoring NA values

# Calculate averages for each week for the LOW data
LOW_avg <- LOW %>%
  group_by(week) %>%
  summarise(average_coverage = mean(coverage, na.rm = TRUE))

# Plot points and add average lines
ggplot() + 
  geom_point(data = HOW, aes(x = week, y = coverage, color = "HOW"), size = 2) +
  geom_point(data = LOW, aes(x = week, y = coverage, color = "LOW"), size = 2) +
  geom_line(data = HOW_avg, aes(x = week, y = average_coverage, color = "HOW"), linewidth = 1) + 
  geom_line(data = LOW_avg, aes(x = week, y = average_coverage, color = "LOW"), linewidth = 1) +
  xlab('week') +
  ylab('coverage') +
  scale_x_continuous(breaks = c(1,2,3,4,5,6,7,8,9,10,11,12,13,14)) +
  scale_color_manual(values = c("HOW" = "red", "LOW" = "blue"),
                     name = "Dataset",  # Legend title
                     labels = c("HOW" = "High initial oxygen", "LOW" = "Low initial oxygen")) +
  geom_vline(xintercept = c(10.5)) +
  my_theme
```
For 31/32, the high initial oxygen samples had very low expression of the homoserine lactone synthetase for weeks 4-10 (oxic conditions), then as soon as the oxygen tension was switched after week 10 (black vertical line), the same samples now had very high increases of expression of this synthetase. For low oxygen conditions, the expression was steady and a bit higher than HO weeks at the same time points, but then peaked at the late weeks 9 and 10, when conditions wer still hypoxic. As soon as the switch happened and these samples had oxygen, the expression immediately dropped and stayed very low. 
 





















Once downloaded and in the correct folder, we need to clean up the files: 
```{r}
# function to load RNAseq file and separate the columns properly
load_and_process_RNAseq_csv <- function(file_paths) {
  processed_dataframes <- list()
  for (filePath in file_paths) {
    data <- read.delim(filePath, header = TRUE, stringsAsFactors = FALSE, row.names = "LOCUS.TAG")
    data <- separate(data, GENE.ID, into = c("STUDY.ID", "ASSEMBLED", "gene_id"), sep = " ", extra = "merge", fill = "right")
    base_name <- tools::file_path_sans_ext(basename(filePath))
    processed_dataframes[[base_name]] <- data
  }
  return(processed_dataframes)
}
```
Want to apply this function to everything in the acidovorax_MTX folder that contains RNAseq/metatranscriptome data for samples aligned to the Acidovorax genome. 
```{r}
# Automatically identifying file paths of RNAseq files in the acidovorax_MT folder
folder_path <- "acidovorax_MTX/" # Ensure this path is correct
file_paths <- list.files(path = folder_path, pattern = "\\.csv$", full.names = TRUE)

# Using the function to load and process the CSV files
processed_dataframes <- load_and_process_RNAseq_csv(file_paths)


# Combine all processed dataframes into one, adding a 'SourceFile' column to each
combined_df <- do.call(rbind, lapply(names(processed_dataframes), function(name) {
  df <- processed_dataframes[[name]]
  df$SourceFile <- name
  return(df)
}))


# to make an individual dataframe
MT33_HOW06_Acidovorax_dfx <- processed_dataframes[["rnaseq_MT33_HOW06_Acidovorax"]]
```

